x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/StringUtils.java

```diff
@@ -327,21 +327,20 @@ public static long parseLong(String string) throws QlIllegalArgumentException {
         }
     }
 
-    public static Number parseInteger(String string) throws QlIllegalArgumentException {
+    public static Number parseIntegral(String string) throws QlIllegalArgumentException {
+        BigInteger bi;
         try {
-            BigInteger bi = new BigInteger(string);
-            try {
-                if (bi.signum() < 0 || bi.compareTo(BigInteger.valueOf(Long.MAX_VALUE)) <= 0) {
-                    return bi.longValueExact();
-                }
-                isUnsignedLong(bi);
-                return bi;
-            } catch (ArithmeticException ae) {
-                throw new QlIllegalArgumentException("Number [{}] is too large", string);
-            }
+            bi = new BigInteger(string);
         } catch (NumberFormatException ex) {
             throw new QlIllegalArgumentException("Cannot parse number [{}]", string);
         }
+        if (bi.signum() < 0 || bi.compareTo(BigInteger.valueOf(Long.MAX_VALUE)) <= 0) {
```

Comment:
> I may be missing something, but why if the sign of the bigInteger is "minus", we are still converting it?

We're converting it to long, in this case.

`parseIntegral()` - called from `ExpressionBuilder` - takes a string as input and tries to resolve it to a long, if possible - i.e. in case the number is negative; or positive, but within long's range - or an unsigned long if not. But since the BigInteger parsing accepts negative numbers, we need to check that explicitly still.

I guess the method could be refactored to only check once if the number is negative, but not sure it would make a relevant positive legibility difference overall.